### ISSUE TITLE
Benchmark

### ISSUE THREAD

**Issue by salvino080-coder** [pytorch/xla#9737]
#!/usr/bin/env python3
"""
Simple benchmark runner for PyTorch/XLA (TPU).
Generates synthetic workloads for common models (resnet50, tiny_mlp) and writes output JSON with timings.

Usage:
  python tools/bench/run_benchmark.py --model resnet50 --batch_size 64 --steps 50 --dtype bf16 --output /tmp/bench.json
"""
import argparse
import json
import time
import torch
import torch.nn as nn
import torch.optim as optim

try:
    import torch_xla.core.xla_model as xm
except Exception:
    xm = None

def get_model(name, device, dtype):
    if name == "resnet50":
        import torchvision.models as models
        m = models.resnet50(pretrained=False)
        m.to(device)
        if dtype == "bf16":
            m = m.to(torch.bfloat16)
        return m
    # tiny MLP for fast smoke
    model = nn.Sequential(
        nn.Flatten(),
        nn.Linear(3*224*224, 4096),
        nn.ReLU(),
        nn.Linear(4096, 1000)
    )
    model.to(device)
    if dtype == "bf16":
        model = model.to(torch.bfloat16)
    return model

def synthetic_loader(batch_size, steps, device, dtype):
    for _ in range(steps):
        x = torch.randn(batch_size, 3, 224, 224)
        y = torch.randint(0, 1000, (batch_size,))
        if dtype == "bf16":
            x = x.to(torch.bfloat16)
        yield x.to(device), y.to(device)

def run_benchmark(args):
    device = xm.xla_device() if xm is not None else torch.device("cpu")
    dtype = args.dtype.lower()
    model = get_model(args.model, device, dtype)
    opt = optim.SGD(model.parameters(), lr=0.01)
    criterion = nn.CrossEntropyLoss()

    times = []
    # warmup steps
    warmup = max(2, int(0.05 * args.steps))
    total_steps = args.steps + warmup
    loader = synthetic_loader(args.batch_size, total_steps, device, dtype)

    for step, (xb, yb) in enumerate(loader, 1):
        start = time.time()
        opt.zero_grad()
        logits = model(xb)
        loss = criterion(logits, yb)
        loss.backward()
        if xm is not None:
            xm.optimizer_step(opt)
            xm.mark_step()
        else:
            opt.step()
        end = time.time()
        if step > warmup:
            times.append(end - start)
        if step >= total_steps:
            break

    # compute metrics
    if len(times) == 0:
        raise RuntimeError("No timed steps collected.")
    step_time_mean = sum(times) / len(times)
    throughput = args.batch_size / step_time_mean
    result = {
        "model": args.model,
        "batch_size": args.batch_size,
        "dtype": args.dtype,
        "steps_measured": len(times),
        "step_time_mean": step_time_mean,
        "throughput_samples_per_sec": throughput,
        "step_time_samples": times[:min(10, len(times))]
    }

    with open(args.output, "w") as f:
        json.dump(result, f, indent=2)
    print("Wrote benchmark results to", args.output)
    print(json.dumps(result, indent=2))

def parse_args():
    p = argparse.ArgumentParser()
    p.add_argument("--model", default="resnet50", choices=["resnet50", "tiny_mlp"])
    p.add_argument("--batch_size", type=int, default=64)
    p.add_argument("--steps", type=int, default=50)
    p.add_argument("--dtype", default="fp32", choices=["fp32", "bf16"])
    p.add_argument("--output", default="/tmp/torch_xla_bench.json")
    return p.parse_args()

if __name__ == "__main__":
    args = parse_args()
    run_benchmark(args)
